File: packages/api-schema/src/types/document.ts
import { z } from 'zod';
import { IdSchema, DateTimeSchema, UserRefSchema, ProjectRefSchema } from './common';
import { DocumentCategoryEnum } from '../enums/status';

/**
 * Document/File schema - Full object
 */
export const DocumentSchema = z.object({
  id: IdSchema,
  name: z.string(),
  type: z.string(), // MIME type or category: image, document, video
  storagePath: z.string(),
  category: DocumentCategoryEnum.nullable(),

  // Drawing-specific fields
  drawingNumber: z.string().nullable(),
  sheetTitle: z.string().nullable(),
  discipline: z.string().nullable(),
  revision: z.string().nullable(),
  scale: z.string().nullable(),

  // Metadata
  pageCount: z.number().int().nullable(),
  currentVersion: z.number().int(),
  isLatest: z.boolean(),
  isAdminOnly: z.boolean(),
  description: z.string().nullable(),
  tags: z.array(z.string()).nullable(),

  // GPS data (for photos)
  gpsLatitude: z.number().nullable(),
  gpsLongitude: z.number().nullable(),
  takenAt: DateTimeSchema.nullable(),

  // Relations
  project: ProjectRefSchema,
  uploader: UserRefSchema,

  // Counts
  revisionCount: z.number().int().optional(),
  annotationCount: z.number().int().optional(),

  // Timestamps
  createdAt: DateTimeSchema,
});
export type Document = z.infer<typeof DocumentSchema>;

/**
 * Document summary (for list views)
 */
export const DocumentSummarySchema = z.object({
  id: IdSchema,
  name: z.string(),
  type: z.string(),
  category: DocumentCategoryEnum.nullable(),
  drawingNumber: z.string().nullable(),
  sheetTitle: z.string().nullable(),
  revisionCount: z.number().int(),
  annotationCount: z.number().int(),
  projectName: z.string().nullable(),
  createdAt: DateTimeSchema,
});
export type DocumentSummary = z.infer<typeof DocumentSummarySchema>;

/**
 * Document revision
 */
export const DocumentRevisionSchema = z.object({
  id: IdSchema,
  version: z.number().int(),
  storagePath: z.string(),
  changeNotes: z.string().nullable(),
  uploadedBy: z.string(),
  fileSize: z.number().int().nullable(),
  createdAt: DateTimeSchema,
});
export type DocumentRevision = z.infer<typeof DocumentRevisionSchema>;

/**
 * Document annotation
 */
export const DocumentAnnotationSchema = z.object({
  id: IdSchema,
  fileId: IdSchema,
  annotationType: z.enum(['COMMENT', 'MARKUP', 'HIGHLIGHT', 'MEASUREMENT', 'CALLOUT']),
  content: z.record(z.unknown()), // Position, text, shape, etc.
  pageNumber: z.number().int().nullable(),
  createdBy: z.string(),
  createdByName: z.string().nullable(),
  resolvedAt: DateTimeSchema.nullable(),
  resolvedBy: z.string().nullable(),
  createdAt: DateTimeSchema,
  updatedAt: DateTimeSchema,
});
export type DocumentAnnotation = z.infer<typeof DocumentAnnotationSchema>;

/**
 * Upload document input
 */
export const UploadDocumentInputSchema = z.object({
  projectId: IdSchema,
  name: z.string().min(1),
  category: DocumentCategoryEnum.optional(),
  description: z.string().nullable().optional(),
  tags: z.array(z.string()).optional(),
  isAdminOnly: z.boolean().optional().default(false),
  dailyLogId: IdSchema.nullable().optional(),

  // GPS for photos
  gpsLatitude: z.number().nullable().optional(),
  gpsLongitude: z.number().nullable().optional(),
});
export type UploadDocumentInput = z.infer<typeof UploadDocumentInputSchema>;

/**
 * Create annotation input
 */
export const CreateAnnotationInputSchema = z.object({
  annotationType: z.enum(['COMMENT', 'MARKUP', 'HIGHLIGHT', 'MEASUREMENT', 'CALLOUT']),
  content: z.record(z.unknown()),
  pageNumber: z.number().int().nullable().optional(),
});
export type CreateAnnotationInput = z.infer<typeof CreateAnnotationInputSchema>;

File: apps/android/construction-android/app/src/main/java/com/constructionpro/app/data/model/Document.kt
package com.constructionpro.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class DocumentsResponse(
  val documents: List<DocumentSummary> = emptyList(),
  val pagination: DocumentPagination? = null,
  val categories: Map<String, Int> = emptyMap()
)

@Serializable
data class DocumentPagination(
  val page: Int? = null,
  val limit: Int? = null,
  val total: Int? = null,
  val pages: Int? = null
)

@Serializable
data class DocumentSummary(
  val id: String,
  val name: String,
  val type: String? = null,
  val category: String? = null,
  val description: String? = null,
  @SerialName("storage_path") val storagePath: String? = null,
  @SerialName("created_at") val createdAt: String? = null,
  val project: DocumentProject? = null,
  val uploader: DocumentUser? = null,
  @SerialName("blaster_assignments") val blasters: List<BlasterAssignment> = emptyList(),
  val metadata: DocumentMetadata? = null,
  val revisions: List<DocumentRevision> = emptyList(),
  @SerialName("_count") val count: DocumentCount? = null
)

@Serializable
data class DocumentProject(
  val id: String? = null,
  val name: String? = null,
  val address: String? = null
)

@Serializable
data class DocumentUser(
  val id: String? = null,
  val name: String? = null,
  val email: String? = null
)

@Serializable
data class BlasterAssignment(
  val id: String,
  val blaster: DocumentUser
)

@Serializable
data class DocumentMetadata(
  val discipline: String? = null,
  @SerialName("drawing_number") val drawingNumber: String? = null,
  @SerialName("sheet_title") val sheetTitle: String? = null,
  val revision: String? = null,
  val scale: String? = null,
  val building: String? = null,
  val floor: String? = null,
  val zone: String? = null
)

@Serializable
data class DocumentRevision(
  val id: String? = null,
  val version: Int? = null,
  @SerialName("storage_path") val storagePath: String? = null,
  @SerialName("change_notes") val changeNotes: String? = null,
  @SerialName("uploaded_by") val uploadedBy: String? = null,
  @SerialName("file_size") val fileSize: Int? = null,
  val checksum: String? = null,
  @SerialName("created_at") val createdAt: String? = null
)

@Serializable
data class DocumentCount(
  val revisions: Int? = null,
  val annotations: Int? = null
)

@Serializable
data class DocumentDetailResponse(
  val document: DocumentDetail
)

@Serializable
data class DocumentDetail(
  val id: String,
  val name: String,
  val type: String? = null,
  val category: String? = null,
  val description: String? = null,
  @SerialName("storage_path") val storagePath: String? = null,
  @SerialName("created_at") val createdAt: String? = null,
  val project: DocumentProject? = null,
  val uploader: DocumentUser? = null,
  val metadata: DocumentMetadata? = null,
  val revisions: List<DocumentRevision> = emptyList(),
  val annotations: List<DocumentAnnotation> = emptyList(),
  @SerialName("_count") val count: DocumentCount? = null
)

@Serializable
data class DocumentAnnotation(
  val id: String? = null,
  @SerialName("annotation_type") val annotationType: String? = null,
  val content: JsonElement? = null,
  @SerialName("page_number") val pageNumber: Int? = null,
  @SerialName("created_by") val createdBy: String? = null,
  @SerialName("resolved_at") val resolvedAt: String? = null,
  @SerialName("resolved_by") val resolvedBy: String? = null,
  @SerialName("created_at") val createdAt: String? = null
)
